Add PhantomData marker to Context to make Context !Send and !Sync

jihiggins opened the pull request:
Adds `PhantomData<*mut ()>` to `Context` in order to allow for future `!Send` or `!Sync` additions to `Context`'s fields. This would allow for things like future single threaded async executor optimizations, or (re)adding `LocalWaker`, etc.

Closes #66481.

Per https://github.com/rust-lang/rust/issues/66481#issuecomment-561289725, this is a breaking change that needs a Crater run as the next step.

(So far have tested the change locally with `cargotest` on WSL)

Commit:
Add PhantomData marker to Context to make Context !Send and !Sync

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

compiler-errors commented:
I believe this is T-libs-api since it affects an (auto) impl of a public type in std

r? rust-lang/libs-api
@rustbot label +T-libs-api

bors commented:
:umbrella: The latest upstream changes (presumably #102022) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

RalfJung commented:
@jihiggins sorry, this seems to have fallen through the cracks. Can you rebase on top of current master and resolve the conflicts? When I can start a crater run.

Cc @rust-lang/wg-async

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3256473616/jobs/5347007991) [(plain)](https://github.com/rust-lang/rust/commit/f4c110319db9db6cd4a353d81c6259ee1e7a6c0a/checks/8910165022/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rand v0.7.3
    Checking std v0.0.0 (/checkout/library/std)
    Checking core v0.0.0 (/checkout/library/core)
    Checking alloc v0.0.0 (/checkout/library/alloc)
error[E0277]: `*mut ()` cannot be shared between threads safely
   |
   |
24 |     static CONTEXT: Context<'static> = Context::from_waker(&WAKER);
   |                     ^^^^^^^^^^^^^^^^ `*mut ()` cannot be shared between threads safely
   |
   = help: within `Context<'static>`, the trait `Sync` is not implemented for `*mut ()`
   = note: required because it appears within the type `PhantomData<*mut ()>`
   = note: required because it appears within the type `Context<'static>`
   = note: shared static variables must have a type that implements `Sync`
For more information about this error, try `rustc --explain E0277`.
error: could not compile `core` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:01:55
```

</details>

RalfJung commented:
Looks like in library/core/tests/task.rs:24:21 we already have a test relying on Context being Sync... that test case can probably be fixed but this is still not a good sign.

@bors try

bors commented:
:hourglass: Trying commit f4c110319db9db6cd4a353d81c6259ee1e7a6c0a with merge 7caa343c25a692951246ea2f0664520b9c3f46da...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f4c110319db9db6cd4a353d81c6259ee1e7a6c0a","merge_sha":"7caa343c25a692951246ea2f0664520b9c3f46da"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3258584088/jobs/5351315534)
Build commit: 7caa343c25a692951246ea2f0664520b9c3f46da (`7caa343c25a692951246ea2f0664520b9c3f46da`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3258584088/jobs/5351315534"},"merge_sha":"7caa343c25a692951246ea2f0664520b9c3f46da"} -->

RalfJung commented:
@craterbot check

Ralith commented:
Reviewing regressions:

Genuine failures:
- Fedomn.rust-knowledge.e2c4007d01b504d4304ec0ed776b26125ffdbf87 is due to futures-async-stream (nightly-only) apparently capturing a `Context` inside a generator
- desync-0.8.1 is an interesting case: [In src/pipe.rs](https://github.com/Logicalshift/desync/blob/cd92466abf5b731c87608517df31e90031e67923/src/pipe.rs#L110-L129) it has a pattern of capturing a context constructed via `from_waker` inside async blocks to later be used when polling an internal future, rather than propagating the context of the enclosing block. I haven't investiggated why this approach is used, but it could conceivably be made compatible by creating `Context`s just in time from captured `Waker`s rather than capturing them directly.

<details><summary>Spurious failures</summary>
Mostly due to failure to determine rustc version at build time, most commonly in libc.

gclark916.vulkano_examples.20dcf2760255f71bf4fdea0ca0f1e9d9eaea221f
gclark916.vulkano_tri1.b3cc87d01ff8e6387c79b8a1942af0d966a6fa74
Ayrx.binja-rs-hello-world.a0f5fb90a3544b233856cddbdac059da21ce27ee
Zoeycode.zoeycode.github.io.74a38dfea8ba84da15b4cb20ce7a38a2fa9ea2e0
etke.bff.2b0c0fb5941f2f751afcdeb15777128afec7e791 
maxco2.pdfwc.5cb4bff8a9f4d3e78b4d64db7788533561c3dce6
stuartZhang.scaffold-wizard.afa3f1ff91917d552310e43422013da28e9f05c5
catboost2-sys-0.1.1+catboost.1.0.5 
catboost2-0.1.1+catboost.1.0.5 
promptconv-0.1.3
ps2-mouse-0.1.4
qt_ritual-0.0.0
readline-async-0.1.0
resctl-bench-2.2.3 
ritual_common-0.4.0 
rmake-0.1.38 
rtsam-0.1.4 
rune-0.12.0 
rusoto_sns-0.48.0 
rust-ini-0.18.0 
rusted_cypher-1.1.0 
rusty-menu-0.1.3 
rvm-0.0.2 
sdoc-0.8.11 
senpy-cli-0.1.1 
serde_pipe-0.1.3 
singletonum-derive-0.2.0 
sjq-crate-publish-demo-0.1.0 
sketch-duplicates-0.1.0 
smafa-0.5.0 
solana-base58-json-converter-0.1.0 
sortnet-0.1.0 
spade-2.0.0 
spglib-sys-1.16.1 
split-csv-0.1.2 
ssandbox-0.1.0 
stable-swap-1.8.1 
stats-cli-3.0.1
</details>

tvallotton commented:
I believe most breakages can be trivially fixed by calling `.waker()` sooner. From what I could tell, it seems to me that the most common case of breakage is due to holding a `Context` across an `.await` boundary.  The worst case scenario is if the handling of `Context`s is part of a crates public API. 

I'm not quite sure what happened in the syn crate. It doesn't seem to be related to this change but if it is it would be concerning. 

For what it's worth, I'd also be willing to submit PRs to the affected crates if that is what it takes for this to be merged.

Ralith commented:
> The worst case scenario is if the handling of Contexts is part of a crates public API.

This does not appear to be the case for desync, fortunately.

tvallotton commented:
Do you reckon, If we made a PR for every broken version of these two crates could this get merged?

jihiggins commented:
How would the order of operations on that work? (Also, what are next steps here? Just reaching out to the maintainers for those crates?)

guswynn commented:
`desync` appears to only be used by crates with the same maintainer: https://crates.io/crates/desync/reverse_dependencies

m-ou-se commented:
Hey @rust-lang/wg-async, we (libs-api) would like your input on this! (There's some context in this issue from 2019: https://github.com/rust-lang/rust/issues/66481)

eholk commented:
We've started a [zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async/topic/.2395985.20-.20Ensure.20.60Context.60.20isnt.20.60Sync.60.2B.60Send.60) on this and also tagged the issue so it will come up in our next triage meeting. We'll report back once we've had the chance to consider it and form some opinions.

tmandry commented:
wg-async is in favor of this change. Context was [never intended](https://www.reddit.com/r/rust/comments/v8e9fa/comment/ibuf6lc) to be Send and Sync. If we can get away with fixing this now, we should do it.

tvallotton commented:
Will this be able to be merged for version `1.65` or will it have to wait for `1.66`?

m-ou-se commented:
@rfcbot merge

dtolnay commented:
I'm marking @yaahc's box because she has stepped down from T-libs-api after the point that this PR got proposed for FCP.

RalfJung commented:
r? libs

dtolnay reviewed (changes requested):
Please fix the test in library/core/tests/task.rs. See https://github.com/rust-lang/rust/pull/95985#issuecomment-1279796220 and https://github.com/rust-lang/rust/pull/95985#issuecomment-1279907314.

dtolnay committed:
Remove test of static Context

Context is no longer Sync so this doesn't work.

    error[E0277]: `*mut ()` cannot be shared between threads safely
      --> library/core/tests/task.rs:24:21
       |
    24 |     static CONTEXT: Context<'static> = Context::from_waker(&WAKER);
       |                     ^^^^^^^^^^^^^^^^ `*mut ()` cannot be shared between threads safely
       |
       = help: within `Context<'static>`, the trait `Sync` is not implemented for `*mut ()`
       = note: required because it appears within the type `PhantomData<*mut ()>`
       = note: required because it appears within the type `Context<'static>`
       = note: shared static variables must have a type that implements `Sync`

dtolnay commented:
I deleted part of the test from #101798 and commented in the tracking issue: https://github.com/rust-lang/rust/issues/102012#issuecomment-1369140411.

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 257e766c0c17c1a3561a509c931cf89bce27dcf5 has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay 257e766c0c17c1a3561a509c931cf89bce27dcf5 -->
<!-- homu: {"type":"Approved","sha":"257e766c0c17c1a3561a509c931cf89bce27dcf5","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jihiggins commented:
> Please fix the test in library/core/tests/task.rs. See [#95985 (comment)](https://github.com/rust-lang/rust/pull/95985#issuecomment-1279796220) and [#95985 (comment)](https://github.com/rust-lang/rust/pull/95985#issuecomment-1279907314).

Sorry, I must've missed those. Do they still need to be fixed, or is that what the subsequent messages are about?

dtolnay commented:
I took care of it in 257e766c0c17c1a3561a509c931cf89bce27dcf5.
